### PR TITLE
Update and extend cufinufft unit tests

### PR DESCRIPTION
**Pull request by janden**
Tests the plan and simple interfaces for various configurations of shapes, number of points, tolerances, and dtypes.

### CODE REVIEW

**Comment by janden** [comment]
Think these are good to go. Let me know if you have any remarks. Will merge by end of week otherwise.

**Review by blackwer — APPROVED** [review]
The whole "assert string in exception message" idiom is strange to me. I guess, unless we define our own exceptions, this is the way. Looks good!

File: python/cufinufft/tests/test_basic.py [hunk 1]
```diff
@@ -7,21 +9,30 @@
 
 import utils
 
+# NOTE: Tests below fail for shapes of size (16, 16) and tolerances 1e-4.
```
Comment (ahbarnett) [inline python/cufinufft/tests/test_basic.py:12]: That's weird. I'd expect more accuracy for smaller N1,N2,N3.   But I checked CPU code and requesting 1e-4 gets as bad as 7e-3 for single-prec (16,16) w/ M=1024. In short, 10.tol is too stringent in general, see below.

File: python/cufinufft/tests/utils.py [hunk 2]
```diff
@@ -37,35 +37,92 @@ def gen_uniform_data(shape, seed=0):
...
+
+    ind = int(0.1789 * np.prod(shape))
+
+    # Ravel the spatial dimensions only.
+    fk_est = fk.reshape(fk.shape[:-dim] + (np.prod(fk.shape[-dim:]),))[..., ind]
+    fk_target = direct_type1(c, k, shape, ind)
+
+    type1_rel_err = np.linalg.norm(fk_target - fk_est) / np.linalg.norm(fk_target)
+
+    assert type1_rel_err < 10 * tol
```
Comment (ahbarnett) [inline python/cufinufft/tests/utils.py:113]: 10.tol cannot always be satisfied. Unless you fix your seed (a good idea) you will get stochastic failure here, for float32 case.   Note in CPU code I use  tol=1e-5 and check for rel l2 err < 2e-4.

File: python/cufinufft/tests/utils.py [hunk 3]
```diff
@@ -37,35 +37,92 @@ def gen_uniform_data(shape, seed=0):
...
+    M = c.shape[-1]
+
+    ind = M // 2
+
+    c_est = c[..., ind]
+    c_target = direct_type2(fk, k[:, ind], dim)
+
+    type2_rel_err = np.linalg.norm(c_target - c_est) / np.linalg.norm(c_target)
+
+    assert type2_rel_err < 10 * tol
```
Comment (ahbarnett) [inline python/cufinufft/tests/utils.py:128]: ditto above.

**Review by ahbarnett — COMMENTED** [review]
Should I (or someone) run these on a cluster GPU - is that needed as part of the review? See tol-pass tweaks below. Thanks, Alex

File: python/cufinufft/tests/test_basic.py [hunk 4]
```diff
@@ -7,21 +9,30 @@
...
-    c = utils.gen_nonuniform_data(M).astype(complex_dtype)
+    k, c = utils.type1_problem(dtype, shape, M)
 
     k_gpu = gpuarray.to_gpu(k)
     c_gpu = gpuarray.to_gpu(c)
 
     plan = Plan(1, shape, eps=tol, dtype=complex_dtype)
 
-    plan.setpts(k_gpu[0], k_gpu[1], k_gpu[2])
+    plan.setpts(*k_gpu)
```
Comment (ahbarnett) [inline python/cufinufft/tests/test_basic.py:35]: has setpts changed to expect a single pointer now (I was expecting kx,ky,kz as before) ?

File: python/cufinufft/tests/test_error_checks.py [hunk 5]
```diff
@@ -64,6 +64,54 @@ def test_set_pts_raises_on_size():
     assert '`z` must be of shape' in err.value.args[0]
 
 
+def test_set_pts_raises_on_nonvector():
```
Comment (ahbarnett) [inline python/cufinufft/tests/test_error_checks.py:67]: I like these analogs of "dumpinputs" for CPU tests.

File: python/cufinufft/tests/test_simple.py [hunk 6]
```diff
@@ -11,77 +11,70 @@
...
-
+def test_simple_type2(dtype, shape, n_trans, M, tol, output_arg):
     real_dtype = dtype
     complex_dtype = utils._complex_dtype(dtype)
 
     dim = len(shape)
 
-    k = utils.gen_nu_pts(M, dim=dim).astype(real_dtype)
-    fk = utils.gen_uniform_data(shape).astype(complex_dtype)
+    fun = {1: cufinufft.nufft1d2,
```
Comment (ahbarnett) [inline python/cufinufft/tests/test_simple.py:63]: this got too abstract for me to understand!

File: python/cufinufft/tests/test_simple.py [hunk 7]
```diff
@@ -11,77 +11,70 @@
...
+
+    k, fk = utils.type2_problem(dtype, shape, M, n_trans=n_trans)
 
     k_gpu = gpuarray.to_gpu(k)
     fk_gpu = gpuarray.to_gpu(fk)
 
     if output_arg:
-        c_gpu = gpuarray.GPUArray((M,), dtype=complex_dtype)
-        cufinufft.nufft3d2(*k_gpu, fk_gpu, out=c_gpu)
+        c_gpu = gpuarray.GPUArray(n_trans + (M,), dtype=complex_dtype)
```
Comment (ahbarnett) [inline python/cufinufft/tests/test_simple.py:73]: likewise as a human I can't follow this ntrans + (M,)   why?    fun(*k_gpu, ... ,...) why?
Maybe just bit more comments for non-pythonistas.

**Comment by janden** [comment]
> The whole "assert string in exception message" idiom is strange to me. I guess, unless we define our own exceptions, this is the way. Looks good!

I've seen it in a few places, although the proper way seems to be to use the `match` kwarg: https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-raises. Will update.

File: python/cufinufft/tests/test_basic.py (hunk 1 above)
Comment (janden) [reply python/cufinufft/tests/test_basic.py:12]: Got it. Relaxed to `25 * tol` and it seems to work fine now (strangely enough, however, the “simple” interface seems to be fine with `10 * tol`).

File: python/cufinufft/tests/utils.py (hunk 2 above)
Comment (janden) [reply python/cufinufft/tests/utils.py:113]: Yep seed is fixed (see data generation in `utils.gen_*_data`).

File: python/cufinufft/tests/test_basic.py (hunk 1 above)
Comment (janden) [reply python/cufinufft/tests/test_basic.py:12]: (So there was a bug in that the simple interface was not setting the tolerance… Fixed now.)

File: python/cufinufft/tests/utils.py (hunk 3 above)
Comment (janden) [reply python/cufinufft/tests/utils.py:128]: Fixed.

File: python/cufinufft/tests/test_basic.py (hunk 4 above)
Comment (janden) [reply python/cufinufft/tests/test_basic.py:35]: So there are a few things going on here. First, the `*` notation acts here to unpack a list (see [here](https://docs.python.org/3.6/tutorial/controlflow.html#unpacking-argument-lists)). The other thing is that multidimensional arrays are treated as lists along their first dimension.

Here we have `k_gpu`, which is an array of shape `(d, M)`, where `d` is the dimension and `M` is the number of points. The above expansion will then yield
```
plan.setpts(k_gpu[0])
```
for `d = 1`,
```
plan.setpts(k_gpu[0], k_gpu[1])
```
for `d = 2`, and
```
plan.setpts(k_gpu[0], k_gpu[1], k_gpu[2])
```
for `d = 3`.

File: python/cufinufft/tests/test_error_checks.py (hunk 5 above)
Comment (janden) [reply python/cufinufft/tests/test_error_checks.py:65]: Yeah I want to make sure that as many lines of code as possible are tested (which includes checking errors, of course).

File: python/cufinufft/tests/test_simple.py (hunk 6 above)
Comment (janden) [reply python/cufinufft/tests/test_simple.py:63]: We're defining a dictionary `{1: nufft1d2, 2: nufft2d1, 3: nufft3d1}` mapping dimensions to function (pointers) and then indexing into it right afterwards `[dim]`. Think of it as a type of switch statement.

File: python/cufinufft/tests/test_simple.py (hunk 7 above)
Comment (janden) [reply python/cufinufft/tests/test_simple.py:73]: So `n_trans` represents the batch shape, which can be `()` (a single transform), `(1,)` (a single transform, but given as an array with first dimension equal to one), or `(n,)` (`n` different transforms).

What this allows us to do is to generate the output array of the correct size by adding the batch shape tuple `n_trans` together with the number of non-uniform points `M`. So in the three above cases we'd get array shapes of `() + (M,) = (M,)`, `(1,) + (M,) = (1, M)`, and `(n,) + (M,) = (n, M),` respectively.

Will add some comments here and in other spots.

**Comment by janden** [comment]
> Should I (or someone) run these on a cluster GPU - is that needed as part of the review?

You can if you'd like. I've run it locally on my laptop, it's run on Jenkins and I just tested it on one of the cluster machines (a V100) and everything looks good.

File: python/cufinufft/tests/test_simple.py (hunk 7 above)
Comment (janden) [reply python/cufinufft/tests/test_simple.py:78]: Done.

**Comment by janden** [comment]
So I think we're good here. Only thing outstanding is that tests still don't run for `tol = 1e-4` (or `1e-5`  or `1e-6` for that matter), which is worrisome. Will require some deeper investigation, but that's not part of the current PR. Will make an issue.

**Pull request merged**